### PR TITLE
Track scroll position

### PR DESCRIPTION
**Pull request by Hixie**
- Change RouteArguments to pass the route's BuildContext rather than
  the Navigator. This caused the bulk of the examples/ and .../test/
  changes (those are mostly mechanical changes).

- Give a key to the StockList. This is what makes it store its scroll
  state between navigations.

- Make initState() actually get called when the State's Element is in
  the tree, so you can use Foo.of() functions there.

- Provide a RouteWidget so that routes have a position in the Widget
  tree. The bulk of the route logic is still in a longer-lived Route
  object for now.

- Make Route.setState() only rebuild the actual route, not the whole
  navigator.

- Provided a Route.of().

- Provided a Route.writeState / Route.readState API that tries to
  identify the clients by their runtimeType, their key, and their
  ancestors keys, up to the nearest ancestor with a GlobalKey.

- Made scrollables hook into this API to track state.

### CODE REVIEW

File: examples/fitness/lib/settings.dart [hunk 1]
```diff
@@ -29,7 +28,7 @@ class SettingsFragmentState extends State<SettingsFragment> {
     return new ToolBar(
       left: new IconButton(
         icon: "navigation/arrow_back",
-        onPressed: config.navigator.pop
+        onPressed: Navigator.of(context).pop
```
Comment (abarth) [inline examples/fitness/lib/settings.dart]: I'd break this out into an anonymous closure so we don't do the walk every build.

File: sky/packages/sky/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -1098,6 +1102,13 @@ class StatefulComponentElement<T extends StatefulComponent, U extends State<T>>
     assert(_state._config == null);
     _state._config = widget;
     assert(_state._debugLifecycleState == _StateLifecycle.created);
+  }
+
+  U get state => _state;
+  U _state;
+
+  void firstBuild() {
```
Comment (abarth) [inline sky/packages/sky/lib/src/widgets/framework.dart]: Can you write a standalone test for this aspect of the change?

File: sky/packages/sky/lib/src/widgets/navigator.dart [hunk 3]
```diff
@@ -330,6 +326,71 @@ class NavigatorState extends State<Navigator> {
...
+  final Route route;
+  final Route nextRoute;
+  final bool buildTargetHeroes;
+  _RouteWidgetState createState() => new _RouteWidgetState();
+}
+
+class _RouteWidgetState extends State<_RouteWidget> {
+  void initState() {
+    super.initState();
+    config.route._widgetState = this;
```
Comment (abarth) [inline sky/packages/sky/lib/src/widgets/navigator.dart]: Why not use a global key to hold onto the state?

File: sky/packages/sky/lib/src/widgets/framework.dart (hunk 2 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/framework.dart]: Will write a test that verifies that initState() is called at a time where you can use 'context' to walk up the tree.

File: sky/packages/sky/lib/src/widgets/navigator.dart [hunk 4]
```diff
@@ -465,6 +532,52 @@ abstract class Route {
...
+      if (element is StatefulComponentElement && element.state is _RouteWidgetState) {
+        result = element.widget.route;
+        return false;
+      }
+      return true;
+    });
+    return result;
+  }
+
+  _StorageEntryIdentifier _computeIdentifier(BuildContext context) {
```
Comment (abarth) [inline sky/packages/sky/lib/src/widgets/navigator.dart]: _computeIdentifier -> _computeStorageIdentifier

File: sky/packages/sky/lib/src/widgets/scrollable.dart [hunk 5]
```diff
@@ -54,15 +55,17 @@ abstract class Scrollable extends StatefulComponent {
 abstract class ScrollableState<T extends Scrollable> extends State<T> {
   void initState() {
     super.initState();
-    if (config.initialScrollOffset is double)
-      _scrollOffset = config.initialScrollOffset;
     _animation = new SimulationStepper(_setScrollOffset);
+    double lastOffset = Route.of(context)?.readState(context);
+    if (lastOffset != null)
```
Comment (abarth) [inline sky/packages/sky/lib/src/widgets/scrollable.dart]: No need for this test.  _scrollOffset should be null already.

**Comment by abarth** [comment]
LGTM but please add a test.

File: sky/packages/sky/lib/src/widgets/navigator.dart (hunk 3 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/navigator.dart]: It would be mildly more expensive. I'm not sure what the advantage would be.

Comment (abarth) [inline sky/packages/sky/lib/src/widgets/navigator.dart]: Don't we already have a global key for each route?  Could use that key for this widget instead of for the keyed subtree?  The advantage would just be simplicity.  Certainly not a big deal either way.

Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/navigator.dart]: I looked at using that key (the subtree key) yesterday, but it's used to insert nodes above the subtree (like an OffStage widget) without losing the render objects.

Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/navigator.dart]: (I think this stuff can certainly be further rationalised, but that's something we should do in a separate patch.)

**Comment by Hixie** [comment]
Added tests.
